src/session/components/message_entry.rs

```diff
@@ -216,6 +228,40 @@ impl MessageEntry {
         self.imp().text_view.buffer().set_text(&text);
     }
 
+    /* formatted_text is being replaced every time text buffer has changed,
+     * so it's very unefficient to parse markdown text when it's not nesessary,
+     * instead we should use a separate method, that will be called only when composing a message */
+
+    pub(crate) async fn parsed_formatted_text(&self) -> Option<FormattedText> {
+        let text = self
+            .imp()
+            .formatted_text
+            .borrow()
+            .clone()
+            .map(|f| f.0)
+            .unwrap_or_default();
+
+        let chat = self.imp().chat.borrow().clone();
+        let client_id = chat.unwrap().session().client_id();
+        functions::parse_markdown(text, client_id)
+            .await
+            .map(|text| {
+                let EnumFormattedText::FormattedText(text) = text;
+                text
+            })
+            .ok()
```

Comment:
I would change the body a bit
```suggestion
        let text = self.imp().formatted_text.borrow().clone().map(|f| f.0)?;
        let client_id = self.chat().unwrap().session().client_id();

        functions::parse_markdown(text.clone(), client_id)
            .await
            .map(|text| {
                let EnumFormattedText::FormattedText(text) = text;
                text
            })
            .ok()
            .or(Some(text))
```
The points the changes addresses:
1. An `?` is used to return early
2. We can use `Self::chat()` method, which you introduced
3. We don't need to create a let binding for chat and clone it
4. Not sure here, but if parsing has failed, we could at least return the original text.

Reply:
Thanks for your help, `or` function existed